File: apps/web/lib/api/modernize.ts
import { API_BASE_URL } from '@/lib/api/endpoints';
import { getAccessToken, refreshAccessToken } from '@/lib/api/client';
import type {
  CodeDiffSummary,
  ModernizeGenerateResponse,
  ModernizeProject,
  ModernizeProjectIngest,
  ModernizeProjectSummary,
  ModernizeResponse,
  RefactorResult,
  RevalidationReport,
} from '@contracts/modernize.contract';
import type { CodebaseAnalysisReport } from '@contracts/codebase-analysis.contract';
import type { ApprovePlanRequest, FixApproval, ModernizationPlan } from '@contracts/refactor-plan.contract';

export type { ModernizeResponse, ModernizeGenerateResponse, IngestStats } from '@contracts/modernize.contract';
export type { CodebaseAnalysisReport, CodeIssue, CodebaseScores } from '@contracts/codebase-analysis.contract';
export type { ModernizationPlan, ModernizationPhase, FixAction, ApprovalMode } from '@contracts/refactor-plan.contract';
export type {
  ModernizeProject,
  ModernizeProjectIngest,
  ModernizeProjectSummary,
  RefactorResult,
  RevalidationReport,
  CodeDiffSummary,
} from '@contracts/modernize.contract';

/** analyze() returns the report + the phased plan together. */
export interface ModernizationReportResponse {
  report: CodebaseAnalysisReport;
  plan: ModernizationPlan;
}

export interface ModernizeAskResponse {
  answer: string;
  mode: 'llm' | 'deterministic';
  grounded_on: string[];
}

export interface RuntimeMetric {
  id: string;
  label: string;
  value: string;
  kind: 'measured' | 'estimate';
  basis: string;
}

export interface RuntimeProfile {
  runtime: string;
  language: string;
  container_ready: boolean;
  executed: boolean;
  metrics: RuntimeMetric[];
  notes: string[];
}

const FIVE_MIN = 5 * 60 * 1000;
const TEN_MIN = 10 * 60 * 1000;

async function send<T>(
  path: string,
  init: RequestInit,
  timeoutMs: number,
  allowRefresh = true,
): Promise<T> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const accessToken = getAccessToken();
    const res = await fetch(`${API_BASE_URL}${path}`, {
      ...init,
      headers: { ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}), ...(init.headers ?? {}) },
      credentials: 'include',
      cache: 'no-store',
      signal: controller.signal,
    });
    if (res.status === 401 && allowRefresh && (await refreshAccessToken())) {
      return send<T>(path, init, timeoutMs, false);
    }
    const text = await res.text();
    const body: unknown = text ? JSON.parse(text) : null;
    if (!res.ok) {
      throw new Error(extractError(body, res.status));
    }
    return body as T;
  } finally {
    clearTimeout(timer);
  }
}

function extractError(body: unknown, status: number): string {
  if (body && typeof body === 'object') {
    const record = body as Record<string, unknown>;
    if (typeof record.detail === 'string') return record.detail;
    if (record.error && typeof record.error === 'object') {
      const message = (record.error as Record<string, unknown>).message;
      if (typeof message === 'string') return message;
    }
  }
  return `HTTP ${status}`;
}

export const modernizeClient = {
  ingestZip: (file: File) => {
    const form = new FormData();
    form.append('file', file);
    return send<ModernizeResponse>('/api/modernize/ingest/zip', { method: 'POST', body: form }, FIVE_MIN);
  },
  ingestGit: (git_url: string) =>
    send<ModernizeResponse>(
      '/api/modernize/ingest/git',
      { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ git_url }) },
      FIVE_MIN,
    ),
  generate: (ingest_id: string, project_name: string, user_model_choice?: string, use_user_key = false) =>
    send<ModernizeGenerateResponse>(
      '/api/modernize/generate',
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ingest_id, project_name, user_model_choice, persist: true, use_user_key }),
      },
      TEN_MIN,
    ),

  // --- Persistent analysis flow (Modernize ingest → Auto-Fix execution) ---
  // Modernize ingests via these; they create a persistent owner-scoped job
  // (project_id == ingest_id) and return the full diagnosis bundle.
  createProjectZip: (file: File) => {
    const form = new FormData();
    form.append('file', file);
    return send<ModernizeProjectIngest>('/api/modernize/projects/upload', { method: 'POST', body: form }, FIVE_MIN);
  },
  createProjectGit: (git_url: string) =>
    send<ModernizeProjectIngest>(
      '/api/modernize/projects/git',
      { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ git_url }) },
      FIVE_MIN,
    ),

  // Auto-Fix picks up an existing analysis — never re-uploads.
  latestProject: () =>
    send<ModernizeProjectSummary | null>('/api/modernize/projects/latest', { method: 'GET' }, FIVE_MIN),
  listProjects: () => send<ModernizeProjectSummary[]>('/api/modernize/projects', { method: 'GET' }, FIVE_MIN),
  deleteProject: (projectId: string) =>
    send<null>(`/api/modernize/projects/${encodeURIComponent(projectId)}`, { method: 'DELETE' }, FIVE_MIN),
  getReport: (projectId: string) =>
    send<ModernizationReportResponse>(`/api/modernize/${projectId}/report`, { method: 'GET' }, FIVE_MIN),
  analyze: (projectId: string, opts: { useUserKey?: boolean; model?: string } = {}) => {
    const params = new URLSearchParams();
    if (opts.useUserKey) params.set('use_user_key', 'true');
    if (opts.model) params.set('user_model_choice', opts.model);
    const qs = params.toString();
    return send<ModernizationReportResponse>(
      `/api/modernize/${projectId}/analyze${qs ? `?${qs}` : ''}`,
      { method: 'POST' },
      TEN_MIN,
    );
  },
  approvePlan: (projectId: string, body: ApprovePlanRequest) =>
    send<FixApproval>(
      `/api/modernize/${projectId}/approve-plan`,
      { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body) },
      FIVE_MIN,
    ),
  applyFixes: (projectId: string) =>
    send<RefactorResult>(`/api/modernize/${projectId}/apply-fixes`, { method: 'POST' }, TEN_MIN),
  revalidate: (projectId: string) =>
    send<RevalidationReport>(`/api/modernize/${projectId}/revalidate`, { method: 'POST' }, FIVE_MIN),
  diff: (projectId: string) => send<CodeDiffSummary>(`/api/modernize/${projectId}/diff`, { method: 'GET' }, FIVE_MIN),

  // Real runtime PROFILE of the ingested codebase (no execution; measured + labeled estimates).
  runtimeProfile: (ingest_id: string) =>
    send<RuntimeProfile>(
      '/api/modernize/runtime-profile',
      { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ ingest_id }) },
      FIVE_MIN,
    ),

  // Grounded Q&A about an ingested project (deterministic without a key; LLM with one).
  ask: (ingest_id: string, question: string, opts: { useUserKey?: boolean; model?: string } = {}) =>
    send<ModernizeAskResponse>(
      '/api/modernize/ask',
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ingest_id, question, use_user_key: opts.useUserKey ?? false, user_model_choice: opts.model }),
      },
      FIVE_MIN,
    ),
};

File: apps/api/app/schemas/modernize.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from app.schemas.common import ApiModel
from app.schemas.generation_validation import GenerationValidationReport

IngestSource = Literal["zip", "git"]
MigrationAction = Literal["migrate", "adapt", "encapsulate", "keep"]
Severity = Literal["info", "low", "medium", "high", "critical"]


class IngestGitRequest(ApiModel):
    git_url: str = Field(min_length=4)


class IngestedFile(ApiModel):
    path: str
    size_bytes: int
    language: str


class CodebaseInventory(ApiModel):
    ingest_id: str
    source: IngestSource
    file_count: int
    total_bytes: int
    skipped_count: int = 0
    languages: dict[str, int] = Field(default_factory=dict)
    files: list[IngestedFile] = Field(default_factory=list)


class SecurityFinding(ApiModel):
    severity: Severity
    code: str
    message: str
    path: str
    line: int | None = None


class ArchitectureSmell(ApiModel):
    code: str
    message: str
    related_paths: list[str] = Field(default_factory=list)


class Diagnosis(ApiModel):
    detected_stack: str
    primary_language: str
    languages: list[str] = Field(default_factory=list)
    dependency_notes: list[str] = Field(default_factory=list)
    smells: list[ArchitectureSmell] = Field(default_factory=list)
    security_findings: list[SecurityFinding] = Field(default_factory=list)


class MigrationMapping(ApiModel):
    legacy_path: str
    target_path: str
    action: MigrationAction
    note: str = ""


class MigrationPlan(ApiModel):
    target_architecture: str
    preserved_logic_note: str
    steps: list[str] = Field(default_factory=list)
    mappings: list[MigrationMapping] = Field(default_factory=list)


class IngestStats(ApiModel):
    """Professional pre-analysis report. Reflects the Smart Ignore Engine: the
    archive may hold hundreds of thousands of files, but only the relevant code is
    ever analyzed. Counts are derived from a single streaming pass."""

    files_found: int = 0  # every file entry in the archive/clone
    ignored_count: int = 0  # auto-ignored (node_modules/.git/build/…) + irrelevant + oversized
    analyzable_count: int = 0  # relevant code files actually indexed for the AI
    total_bytes: int = 0  # effective analyzable code size (uncompressed)
    lines_of_code: int = 0
    languages: dict[str, int] = Field(default_factory=dict)
    frameworks: list[str] = Field(default_factory=list)
    complexity: str = "unknown"
    truncated: bool = False  # analyzable cap reached; remaining relevant files were skipped

class ModernizeScoreMetric(ApiModel):
    id: str
    label: str
    value: int = Field(ge=0, le=100)
    basis: str


class ModernizeFindingSummary(ApiModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    info: int = 0
    total: int = 0
    technical_debt: int = 0
    duplicated_code: int = 0
    dead_code: int = 0
    dependencies: int = 0


class ModernizeDetectedTechnology(ApiModel):
    name: str
    category: str
    confidence: int = Field(ge=0, le=100)
    evidence: str


class ModernizeExecutiveSummary(ApiModel):
    overall_health: int = Field(ge=0, le=100)
    health_label: str
    modernization_estimate: str
    complexity: str
    risk_level: Literal["low", "medium", "high"]
    analysis_confidence: int = Field(ge=0, le=100)
    scores: list[ModernizeScoreMetric] = Field(default_factory=list)
    findings: ModernizeFindingSummary = Field(default_factory=ModernizeFindingSummary)
    technologies: list[ModernizeDetectedTechnology] = Field(default_factory=list)
    review: str
    priority: str

class ModernizeResponse(ApiModel):
    inventory: CodebaseInventory
    diagnosis: Diagnosis
    plan: MigrationPlan
    stats: IngestStats | None = None
    executive_summary: ModernizeExecutiveSummary | None = None


class ModernizeGenerateRequest(ApiModel):
    ingest_id: str = Field(min_length=1)
    project_name: str = "modernized-project"
    user_model_choice: str | None = None
    persist: bool = True
    use_user_key: bool = False


class ModernizeDeepAnalyzeRequest(ApiModel):
    ingest_id: str = Field(min_length=1)
    pace: bool = True  # deliberate per-stage pacing (disable for tests)


class ModernizeAskRequest(ApiModel):
    ingest_id: str = Field(min_length=1)
    question: str = Field(min_length=1, max_length=2000)
    use_user_key: bool = False
    user_model_choice: str | None = None


class ModernizeAskResponse(ApiModel):
    answer: str
    mode: Literal["llm", "deterministic"]
    grounded_on: list[str] = Field(default_factory=list)


class ModernizeRuntimeRequest(ApiModel):
    ingest_id: str = Field(min_length=1)


class RuntimeMetric(ApiModel):
    id: str
    label: str
    value: str
    # measured = computed from the real code; estimate = labeled heuristic (basis stated).
    kind: Literal["measured", "estimate"]
    basis: str


class RuntimeProfile(ApiModel):
    contractVersion: str
    runtime: str
    language: str
    container_ready: bool
    executed: bool = False  # the legacy code is NEVER executed; this stays False by design
    metrics: list[RuntimeMetric] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class ModernizeGenerateResponse(ApiModel):
    ok: bool
    project_id: str | None = None
    file_count: int = 0
    written: bool = False
    degraded: bool = False
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    validation_report: GenerationValidationReport | None = None


# --------------------------------------------------------------------------- #
# Modernize pipeline (deep analysis + approved auto-refactor)
# --------------------------------------------------------------------------- #

IssueSeverity = Literal["info", "low", "medium", "high", "critical"]
RiskLevel = Literal["low", "medium", "high"]
ApprovalMode = Literal["critical_only", "full", "custom"]


class ModernizeProject(ApiModel):
    project_id: str
    source: IngestSource
    file_count: int
    total_bytes: int
    skipped_count: int = 0
    languages: dict[str, int] = Field(default_factory=dict)
    created_at: str
    stats: IngestStats | None = None


class ModernizeProjectIngest(ModernizeResponse):
    """Persistent-job ingest result: the full diagnosis bundle the cockpit renders
    (inventory/diagnosis/plan/stats/executive_summary) PLUS the persistent
    project_id (== ingest_id) so Modernize can hand the analysis to Auto-Fix.
    Carries the flat project fields too for backward compatibility with the
    previous ModernizeProject response shape."""

    project_id: str
    source: IngestSource
    created_at: str
    file_count: int = 0
    total_bytes: int = 0
    skipped_count: int = 0
    languages: dict[str, int] = Field(default_factory=dict)


class LlmProviderConfig(ApiModel):
    id: Literal["openai", "anthropic", "google", "deepseek", "groq"]
    name: str
    description: str
    recommended_for: str
    key_required: bool = True
    status: Literal["ready", "initializing", "not_configured", "auth_error", "unavailable"] = "not_configured"


class LlmProviderCatalog(ApiModel):
    providers: list[LlmProviderConfig] = Field(default_factory=list)


class LlmConnectionTestRequest(ApiModel):
    provider: str


class LlmConnectionTestResult(ApiModel):
    ok: bool
    provider: str
    model: str | None = None
    # Human message; NEVER contains the key.
    message: str
    degraded: bool = False


class CodebaseScores(ApiModel):
    architecture: int = 0
    security: int = 0
    backend: int = 0
    frontend: int = 0
    database: int = 0
    tests: int = 0
    devops: int = 0
    maintainability: int = 0
    performance: int = 0
    production_readiness: int = 0
    overall: int = 0


class CodeIssue(ApiModel):
    id: str
    title: str
    severity: IssueSeverity
    category: str
    file: str | None = None
    line: int | None = None
    root_cause: str
    recommendation: str
    auto_fixable: bool = False


class ExecutiveReport(ApiModel):
    health: str
    risk_level: RiskLevel
    top_problems: list[str] = Field(default_factory=list)
    business_impact: str
    effort_estimate: str
    priority: str


class TechnicalReport(ApiModel):
    issues: list[CodeIssue] = Field(default_factory=list)


class CodebaseAnalysisReport(ApiModel):
    project_id: str
    detected_stack: str
    primary_language: str
    scores: CodebaseScores
    executive: ExecutiveReport
    technical: TechnicalReport
    degraded: bool = False  # True when no real LLM enriched the analysis
    generated_at: str


class FixAction(ApiModel):
    id: str  # auto-repair issue id when auto_fixable
    title: str
    phase_id: str
    auto_fixable: bool = False
    requires_extra_confirmation: bool = False


class ModernizationPhase(ApiModel):
    id: str  # critical | security | architecture | tests | devops
    title: str
    actions: list[FixAction] = Field(default_factory=list)


class ModernizationPlan(ApiModel):
    project_id: str
    phases: list[ModernizationPhase] = Field(default_factory=list)


class ModernizationReportResponse(ApiModel):
    report: CodebaseAnalysisReport
    plan: ModernizationPlan


class ModernizeProjectSummary(ApiModel):
    """Lightweight card for picking up an existing analysis (latest / list).
    Carries enough to render the Auto-Fix header + empty/loading states in one
    round-trip; the full report is fetched separately via /report."""

    project_id: str
    source: IngestSource
    file_count: int
    total_bytes: int = 0
    languages: dict[str, int] = Field(default_factory=dict)
    created_at: str
    updated_at: str
    has_report: bool = False
    detected_stack: str | None = None
    overall_score: int | None = None
    findings_count: int | None = None
    scores: CodebaseScores | None = None


class ApprovePlanRequest(ApiModel):
    mode: ApprovalMode = "critical_only"
    phase_ids: list[str] = Field(default_factory=list)  # used when mode == "custom"


class FixApproval(ApiModel):
    project_id: str
    approved_phase_ids: list[str] = Field(default_factory=list)


class RefactorResult(ApiModel):
    project_id: str
    materialized_project_id: str
    applied_count: int = 0
    failed_count: int = 0
    actions: list[str] = Field(default_factory=list)  # human action titles
    diff_summary: list[str] = Field(default_factory=list)  # paths only


class RevalidationReport(ApiModel):
    project_id: str
    before: CodebaseScores
    after: CodebaseScores
    deltas: dict[str, int] = Field(default_factory=dict)
    degraded: bool = False


class CodeDiffSummary(ApiModel):
    project_id: str
    changed_paths: list[str] = Field(default_factory=list)
    score_deltas: dict[str, int] = Field(default_factory=dict)


class ModernizeFlags(ApiModel):
    modernize_enabled: bool
    modernize_git_import: bool
    modernize_zip_upload: bool
    modernize_auto_refactor: bool
    modernize_user_llm_key: bool
    modernize_export: bool


class ModernizeConfigResponse(ApiModel):
    flags: ModernizeFlags
    providers: list[LlmProviderConfig] = Field(default_factory=list)
